# Learning Points & Summary
In this lesson we looked at how to:

* Pull a random sample from a DataFrame using `.sample()`

* How to find duplicate entries with `.duplicated()` and `.drop_duplicates()`

* How to convert string and object data types into numbers with `.to_numeric()`

* How to use plotly to generate beautiful pie, donut, and bar charts as well as box and scatter plots

# Introduction

In this notebook, we will do a comprehensive analysis of the Android app market by comparing thousands of apps in the Google Play store.

# About the Dataset of Google Play Store Apps & Reviews

**Data Source:** <br>
App and review data was scraped from the Google Play Store by Lavanya Gupta in 2018. Original files listed [here](
https://www.kaggle.com/lava18/google-play-store-apps).

# Import Statements

In [148]:
import pandas as pd
import plotly.express as px

# Notebook Presentation

In [149]:
# Show numeric output in decimal format e.g., 2.15
pd.options.display.float_format = '{:,.2f}'.format

# Read the Dataset

In [150]:
df_apps = pd.read_csv('apps.csv')

# Data Cleaning

**Challenge**: How many rows and columns does `df_apps` have? What are the column names? Look at a random sample of 5 different rows with [.sample()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html).

In [151]:
df_apps.shape # 10841 rows and 12 columns

(10841, 12)

In [152]:
# Column names
df_apps.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size_MBs', 'Installs', 'Type',
       'Price', 'Content_Rating', 'Genres', 'Last_Updated', 'Android_Ver'],
      dtype='object')

In [153]:
# 5 random rows:
df_apps.sample(5)

App  ...   Android_Ver
10643                                    Angry Birds Rio  ...    4.1 and up
10560             OfficeSuite : Free Office + PDF Editor  ...    4.1 and up
2164                        Business Intelligence & Data  ...    2.0 and up
5584          Gay Sugar Daddy Dating & Hookup – Sudy Gay  ...  4.0.3 and up
8596   L.POINT - 엘포인트 [ 포인트, 멤버십, 적립, 사용, 모바일 카드, 쿠폰,...  ...    4.1 and up

[5 rows x 12 columns]

### Drop Unused Columns

**Challenge**: Remove the columns called `Last_Updated` and `Android_Version` from the DataFrame. We will not use these columns. 

In [154]:
# Setting axis=1 to specify that we want to drop certain columns:
df_apps_clean = df_apps.drop(['Last_Updated', 'Android_Ver'], axis=1, inplace=True)

In [155]:
df_apps.head()

App  ...           Genres
0                       Ak Parti Yardım Toplama  ...           Social
1                    Ain Arabic Kids Alif Ba ta  ...        Education
2  Popsicle Launcher for Android P 9.0 launcher  ...  Personalization
3                     Command & Conquer: Rivals  ...         Strategy
4                                    CX Network  ...         Business

[5 rows x 10 columns]

### Find and Remove NaN values in Ratings

**Challenge**: How may rows have a NaN value (not-a-number) in the Ratings column? Create DataFrame called `df_apps_clean` that does not include these rows. 

In [156]:
nan_rows = df_apps[df_apps.Rating.isna()]
print(nan_rows.shape)
nan_rows.head()

(1474, 10)


App  ...           Genres
0                       Ak Parti Yardım Toplama  ...           Social
1                    Ain Arabic Kids Alif Ba ta  ...        Education
2  Popsicle Launcher for Android P 9.0 launcher  ...  Personalization
3                     Command & Conquer: Rivals  ...         Strategy
4                                    CX Network  ...         Business

[5 rows x 10 columns]

In [157]:
df_apps_clean = df_apps.dropna()

In [158]:
df_apps_clean.shape

(9367, 10)

### Find and Remove Duplicates

**Challenge**: Are there any duplicates in data? Check for duplicates using the [.duplicated()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) function. How many entries can you find for the "Instagram" app? Use [.drop_duplicates()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) to remove any duplicates from `df_apps_clean`. 


In [159]:
duplicated_rows = df_apps_clean[df_apps_clean.duplicated()]
print(duplicated_rows.shape)
duplicated_rows.head()

(476, 10)


App  ...   Genres
946                       420 BZ Budeze Delivery  ...  Medical
1133                                 MouseMingle  ...   Dating
1196  Cardiac diagnosis (heart rate, arrhythmia)  ...  Medical
1231                                Sway Medical  ...  Medical
1247              Chat Kids - Chat Room For Kids  ...   Dating

[5 rows x 10 columns]

In [160]:
df_apps_clean[df_apps_clean.App == 'Instagram']

App Category  Rating   Reviews  ...  Type Price Content_Rating  Genres
10806  Instagram   SOCIAL    4.50  66577313  ...  Free     0           Teen  Social
10808  Instagram   SOCIAL    4.50  66577446  ...  Free     0           Teen  Social
10809  Instagram   SOCIAL    4.50  66577313  ...  Free     0           Teen  Social
10810  Instagram   SOCIAL    4.50  66509917  ...  Free     0           Teen  Social

[4 rows x 10 columns]

In [161]:
df_apps_clean = df_apps_clean.drop_duplicates()
df_apps_clean[df_apps_clean.App == 'Instagram']

App Category  Rating   Reviews  ...  Type Price Content_Rating  Genres
10806  Instagram   SOCIAL    4.50  66577313  ...  Free     0           Teen  Social
10808  Instagram   SOCIAL    4.50  66577446  ...  Free     0           Teen  Social
10810  Instagram   SOCIAL    4.50  66509917  ...  Free     0           Teen  Social

[3 rows x 10 columns]

In [162]:
# Need to specify the subset columns used in the comparision to identify duplicates:
df_apps_clean = df_apps_clean.drop_duplicates(subset=['App', 'Type', 'Price'])
df_apps_clean[df_apps_clean.App == 'Instagram']

App Category  Rating   Reviews  ...  Type Price Content_Rating  Genres
10806  Instagram   SOCIAL    4.50  66577313  ...  Free     0           Teen  Social

[1 rows x 10 columns]

# Find Highest Rated Apps

**Challenge**: Identify which apps are the highest rated. What problem might you encounter if you rely exclusively on ratings alone to determine the quality of an app?

In [163]:
df_apps_clean.head(1)
df_apps_clean.sort_values('Rating', ascending=False).head()

App     Category  ...  Content_Rating       Genres
21    KBA-EZ Health Guide      MEDICAL  ...        Everyone      Medical
1230         Sway Medical      MEDICAL  ...        Everyone      Medical
1227    AJ Men's Grooming    LIFESTYLE  ...        Everyone    Lifestyle
1224       FK Dedinje BGD       SPORTS  ...        Everyone       Sports
1223      CB VIDEO VISION  PHOTOGRAPHY  ...        Everyone  Photography

[5 rows x 10 columns]

In [164]:
df_apps_clean[df_apps_clean.Rating == 5.00].sort_values('Reviews', ascending=False)
# 271 apps with the 5* rating

App  ...               Genres
2095                                         Ríos de Fe  ...            Lifestyle
2438  FD Calculator (EMI, SIP, RD & Loan Eligilibility)  ...              Finance
3115                                         Oración CX  ...            Lifestyle
2107                     Barisal University App-BU Face  ...            Education
2069                                         Master E.K  ...            Education
...                                                 ...  ...                  ...
1272                                          CG Prints  ...          Photography
453                                 Wowkwis aq Ka'qaquj  ...  Education;Education
349                                       DF Glue Board  ...            Parenting
230   NCLEX Multi-topic Nursing Exam Review-Quiz & n...  ...              Medical
411                                           EC SPORTS  ...               Sports

[271 rows x 10 columns]

# Find 5 Largest Apps in terms of Size (MBs)

**Challenge**: What's the size in megabytes (MB) of the largest Android apps in the Google Play Store. Based on the data, do you think there could be limit in place or can developers make apps as large as they please? 

In [165]:
df_apps_clean.sort_values('Size_MBs', ascending=False).head()
df_apps_clean[df_apps_clean.Size_MBs == 100].shape # 14 apps with size of 100MB

(14, 10)

# Find the 5 App with Most Reviews

**Challenge**: Which apps have the highest number of reviews? Are there any paid apps among the top 50?

In [166]:
df_apps_clean.sort_values('Reviews', ascending=False).head()

App  ...         Genres
10805                                  Facebook  ...         Social
10785                        WhatsApp Messenger  ...  Communication
10806                                 Instagram  ...         Social
10784  Messenger – Text and Video Chat for Free  ...  Communication
10650                            Clash of Clans  ...       Strategy

[5 rows x 10 columns]

In [167]:
top_50 = df_apps_clean.sort_values('Reviews', ascending=False).head(50)
top_50[top_50.Type == 'Paid'].count()

App               0
Category          0
Rating            0
Reviews           0
Size_MBs          0
Installs          0
Type              0
Price             0
Content_Rating    0
Genres            0
dtype: int64

# Plotly Pie and Donut Charts - Visualise Categorical Data: Content Ratings

In [168]:
ratings = df_apps_clean.Content_Rating.value_counts()
ratings

Everyone           6621
Teen                912
Mature 17+          357
Everyone 10+        305
Adults only 18+       3
Unrated               1
Name: Content_Rating, dtype: int64

In [169]:
# Create a graph_object:
fig = px.pie(labels=ratings.index, values=ratings.values)

fig.show()

In [170]:
fig = px.pie(labels=ratings.index, values=ratings.values,
             title="Content Rating", names=ratings.index)

fig.update_traces(textposition='outside', textinfo='percent+label')

fig.show()

In [171]:
# Add 'hole' to create a donut 🍩 chart:
fig = px.pie(labels=ratings.index,
             values=ratings.values,
             title="Content Rating",
             names=ratings.index,
             hole=0.6
             )

fig.update_traces(textposition='inside', 
                  textfont_size=15, 
                  textinfo='percent')
 
fig.show()

In [172]:
fig = px.pie(df_apps_clean, names='Content_Rating')
fig.show()

# Numeric Type Conversion: Examine the Number of Installs

**Challenge**: How many apps had over 1 billion (that's right - BILLION) installations? How many apps just had a single install? 

Check the datatype of the Installs column.

Count the number of apps at each level of installations. 

Convert the number of installations (the Installs column) to a numeric data type. Hint: this is a 2-step process. You'll have make sure you remove non-numeric characters first. 

In [173]:
df_apps_clean.Installs.describe()

count          8199
unique           19
top       1,000,000
freq           1417
Name: Installs, dtype: object

In [174]:
df_apps_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8199 entries, 21 to 10835
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             8199 non-null   object 
 1   Category        8199 non-null   object 
 2   Rating          8199 non-null   float64
 3   Reviews         8199 non-null   int64  
 4   Size_MBs        8199 non-null   float64
 5   Installs        8199 non-null   object 
 6   Type            8199 non-null   object 
 7   Price           8199 non-null   object 
 8   Content_Rating  8199 non-null   object 
 9   Genres          8199 non-null   object 
dtypes: float64(2), int64(1), object(7)
memory usage: 704.6+ KB


In [175]:
df_apps_clean.dtypes

App                object
Category           object
Rating            float64
Reviews             int64
Size_MBs          float64
Installs           object
Type               object
Price              object
Content_Rating     object
Genres             object
dtype: object

In [176]:
df_apps_clean.Installs.value_counts()

1,000,000        1417
100,000          1096
10,000            988
10,000,000        933
1,000             698
5,000,000         607
500,000           504
50,000            457
5,000             425
100               303
50,000,000        202
500               199
100,000,000       189
10                 69
50                 56
500,000,000        24
1,000,000,000      20
5                   9
1                   3
Name: Installs, dtype: int64

In [177]:
# Remove all the commas in the Installs column:
df_apps_clean.Installs = df_apps_clean.Installs.astype(str).str.replace(',','')

# Convert string to number:
df_apps_clean.Installs = pd.to_numeric(df_apps_clean.Installs)

df_apps_clean.Installs.value_counts()

# Count the number of entries per level of installations
df_apps_clean[['App', 'Installs']].groupby('Installs').count()

App
Installs        
1              3
5              9
10            69
50            56
100          303
500          199
1000         698
5000         425
10000        988
50000        457
100000      1096
500000       504
1000000     1417
5000000      607
10000000     933
50000000     202
100000000    189
500000000     24
1000000000    20

# Find the Most Expensive Apps, Filter out the Junk, and Calculate a (ballpark) Sales Revenue Estimate

Let's examine the Price column more closely.

**Challenge**: Convert the price column to numeric data. Then investigate the top 20 most expensive apps in the dataset.

Remove all apps that cost more than $250 from the `df_apps_clean` DataFrame.

Add a column called 'Revenue_Estimate' to the DataFrame. This column should hold the price of the app times the number of installs. What are the top 10 highest grossing paid apps according to this estimate? Out of the top 10 highest grossing paid apps, how many are games?


In [178]:
# Convert the price column to numeric data:
df_apps_clean.Price = df_apps_clean.Price.astype(str).str.replace('$','')
df_apps_clean.Price = pd.to_numeric(df_apps_clean.Price)

df_apps_clean[['App', 'Price']].head(3)

App  Price
21  KBA-EZ Health Guide   0.00
28             Ra Ga Ba   1.49
47              Mu.F.O.   0.99

In [179]:
# Top 20 most expensive apps in the dataset:
top_20_price = df_apps_clean.sort_values('Price', ascending=False).head(20)
top_20_price

App   Category  ...  Content_Rating         Genres
3946        I'm Rich - Trump Edition  LIFESTYLE  ...        Everyone      Lifestyle
2461              I AM RICH PRO PLUS    FINANCE  ...        Everyone        Finance
4606               I Am Rich Premium    FINANCE  ...        Everyone        Finance
3145              I am rich(premium)    FINANCE  ...        Everyone        Finance
3554                      💎 I'm rich  LIFESTYLE  ...        Everyone      Lifestyle
5765                       I am rich  LIFESTYLE  ...        Everyone      Lifestyle
1946  I am rich (Most expensive app)    FINANCE  ...            Teen        Finance
2775                   I Am Rich Pro     FAMILY  ...        Everyone  Entertainment
3221                  I am Rich Plus     FAMILY  ...        Everyone  Entertainment
3114                       I am Rich    FINANCE  ...        Everyone        Finance
1331          most expensive app (H)     FAMILY  ...        Everyone  Entertainment
2394                      I am Rich!    FINANCE  ...        Everyone        Finance
3897                       I Am Rich     FAMILY  ...        Everyone  Entertainment
2193             I am extremely Rich  LIFESTYLE  ...        Everyone      Lifestyle
3856                   I am rich VIP  LIFESTYLE  ...        Everyone      Lifestyle
2281       Vargo Anesthesia Mega App    MEDICAL  ...        Everyone        Medical
1407                    LTC AS Legal    MEDICAL  ...        Everyone        Medical
2629                I am Rich Person  LIFESTYLE  ...        Everyone      Lifestyle
2481         A Manual of Acupuncture    MEDICAL  ...        Everyone        Medical
4264         Golfshot Plus: Golf GPS     SPORTS  ...        Everyone         Sports

[20 rows x 10 columns]

### The most expensive apps sub $250

In [180]:
df_apps_clean = df_apps_clean[df_apps_clean.Price < 250]
df_apps_clean.sort_values('Price', ascending=False).head(10)

App  ...     Genres
2281                          Vargo Anesthesia Mega App  ...    Medical
1407                                       LTC AS Legal  ...    Medical
2629                                   I am Rich Person  ...  Lifestyle
2481                            A Manual of Acupuncture  ...    Medical
2463                                 PTA Content Master  ...    Medical
2207                                           EMT PASS  ...    Medical
4264                            Golfshot Plus: Golf GPS  ...     Sports
504                           AP Art History Flashcards  ...  Education
4772   Human Anatomy Atlas 2018: Complete 3D Human Body  ...    Medical
3241  Muscle Premium - Human Anatomy, Kinesiology, B...  ...    Medical

[10 rows x 10 columns]

### Highest Grossing Paid Apps (ballpark estimate)

In [181]:
df_apps_clean['Revenue_Estimate'] = df_apps_clean.Installs.mul(df_apps_clean.Price)
df_apps_clean.sort_values('Revenue_Estimate', ascending=False)[:10]

App  ... Revenue_Estimate
9220                      Minecraft  ...    69,900,000.00
8825                  Hitman Sniper  ...     9,900,000.00
7151  Grand Theft Auto: San Andreas  ...     6,990,000.00
7477            Facetune - For Free  ...     5,990,000.00
7977        Sleep as Android Unlock  ...     5,990,000.00
6594            DraStic DS Emulator  ...     4,990,000.00
6082                   Weather Live  ...     2,995,000.00
7954                    Bloons TD 5  ...     2,990,000.00
7633        Five Nights at Freddy's  ...     2,990,000.00
6746     Card Wars - Adventure Time  ...     2,990,000.00

[10 rows x 11 columns]

# Plotly Bar Charts & Scatter Plots: Analysing App Categories

### Vertical Bar Chart - Highest Competition (Number of Apps)

In [182]:
# Number of different categories
df_apps_clean.Category.nunique()

33

In [183]:
top10_category = df_apps_clean.Category.value_counts()[:10]
top10_category

FAMILY             1606
GAME                910
TOOLS               719
PRODUCTIVITY        301
PERSONALIZATION     298
LIFESTYLE           297
FINANCE             296
MEDICAL             292
PHOTOGRAPHY         263
BUSINESS            262
Name: Category, dtype: int64

In [184]:
bar = px.bar(x=top10_category.index,
             y=top10_category.values)

bar.show()

### Horizontal Bar Chart - Most Popular Categories (Highest Downloads)

In [185]:
# Group all our apps by category and sum the number of installations:
category_installs = df_apps_clean.groupby('Category').agg({'Installs': pd.Series.sum})
category_installs.sort_values('Installs', ascending=True, inplace=True)

category_installs

Installs
Category                        
EVENTS                  15949410
BEAUTY                  26916200
PARENTING               31116110
MEDICAL                 39162676
COMICS                  44931100
LIBRARIES_AND_DEMO      52083000
AUTO_AND_VEHICLES       53129800
HOUSE_AND_HOME          97082000
ART_AND_DESIGN         114233100
DATING                 140912410
FOOD_AND_DRINK         211677750
EDUCATION              352852000
WEATHER                361096500
FINANCE                455249400
MAPS_AND_NAVIGATION    503267560
LIFESTYLE              503611120
BUSINESS               692018120
SPORTS                1096431465
HEALTH_AND_FITNESS    1134006220
SHOPPING              1400331540
PERSONALIZATION       1532352930
BOOKS_AND_REFERENCE   1665791655
ENTERTAINMENT         2113660000
NEWS_AND_MAGAZINES    2369110650
TRAVEL_AND_LOCAL      2894859300
VIDEO_PLAYERS         3916897200
FAMILY                4437554490
PHOTOGRAPHY           4649143130
SOCIAL                5487841475
PRODUCTIVITY          5788070180
TOOLS                 8099724500
COMMUNICATION        11039241530
GAME                 13858762717

In [186]:
# Create horizonal bar:
h_bar = px.bar(x=category_installs.Installs,
               y=category_installs.index,
               orientation='h',
               title='Category Popularity',
               height=800)

h_bar.update_layout(xaxis_title='Number of Downloads',
                    yaxis_title='Category')

h_bar.show()

### Category Concentration - Downloads vs. Competition

**Challenge**: 
* First, create a DataFrame that has the number of apps in one column and the number of installs in another:

<img src=https://imgur.com/uQRSlXi.png width="350">

* Then use the [plotly express examples from the documentation](https://plotly.com/python/line-and-scatter/) alongside the [.scatter() API reference](https://plotly.com/python-api-reference/generated/plotly.express.scatter.html)to create scatter plot that looks like this. 

<img src=https://imgur.com/cHsqh6a.png>

*Hint*: Use the size, hover_name and color parameters in `.scatter()`. To scale the yaxis, call `.update_layout()` and specify that the yaxis should be on a log-scale like so: `yaxis=dict(type='log')`

In [187]:
category_concentration = df_apps_clean.groupby('Category').agg({'App':pd.Series.count,'Installs':pd.Series.sum})

print(category_concentration.shape)
category_concentration.sort_values('Installs', ascending=False)

(33, 2)


App     Installs
Category                              
GAME                  910  13858762717
COMMUNICATION         257  11039241530
TOOLS                 719   8099724500
PRODUCTIVITY          301   5788070180
SOCIAL                203   5487841475
PHOTOGRAPHY           263   4649143130
FAMILY               1606   4437554490
VIDEO_PLAYERS         148   3916897200
TRAVEL_AND_LOCAL      187   2894859300
NEWS_AND_MAGAZINES    204   2369110650
ENTERTAINMENT         102   2113660000
BOOKS_AND_REFERENCE   169   1665791655
PERSONALIZATION       298   1532352930
SHOPPING              180   1400331540
HEALTH_AND_FITNESS    243   1134006220
SPORTS                260   1096431465
BUSINESS              262    692018120
LIFESTYLE             297    503611120
MAPS_AND_NAVIGATION   118    503267560
FINANCE               296    455249400
WEATHER                72    361096500
EDUCATION             118    352852000
FOOD_AND_DRINK         94    211677750
DATING                134    140912410
ART_AND_DESIGN         61    114233100
HOUSE_AND_HOME         62     97082000
AUTO_AND_VEHICLES      73     53129800
LIBRARIES_AND_DEMO     64     52083000
COMICS                 54     44931100
MEDICAL               292     39162676
PARENTING              50     31116110
BEAUTY                 42     26916200
EVENTS                 45     15949410

In [188]:
scatter = px.scatter(category_concentration, 
                     x='App',
                     y='Installs',
                     title='Category Concentration',
                     size='App', # size of the buble
                     hover_name=category_concentration.index,
                     color='Installs') # rainbow colors 🏳️‍🌈

scatter.update_layout(xaxis_title='Number of Apps (Lower=More Concentrated)',
                      yaxis_title='Installs',
                      yaxis=dict(type='log')) # log-scale

scatter.show()

What we see is that the categories like Family, Tools, and Game have many different apps sharing a high number of downloads. But for the categories like video players and entertainment, all the downloads are concentrated in very few apps.
<img src='https://img-c.udemycdn.com/redactor/raw/2020-10-11_13-16-11-a310e773b06e1efc0ad4114a12a51e01.png'>

### Installs/App column:

In [189]:
cat_con = df_apps_clean.groupby('Category').agg({'App':pd.Series.count,'Installs':pd.Series.sum})
cat_con.sort_values('Installs', ascending=False, inplace=True)
cat_con['Installs/App'] = cat_con.Installs.div(cat_con.App)

scatter = px.scatter(cat_con,
                     x='App',
                     y='Installs',
                     log_y=True, # ???
                     color='Installs/App',
                     color_continuous_scale=px.colors.sequential.Plasma_r,
                     size=cat_con.App,
                     size_max=40,
                     hover_name=cat_con.index,
                     title='Category Concentration'
                     )

scatter.update_layout(xaxis_title='Number of Apps (Lower=More Concentrated)',
                      yaxis_title='Installs')

scatter.show()

# Extracting Nested Data from a Column

**Challenge**: How many different types of genres are there? Can an app belong to more than one genre? Check what happens when you use `.value_counts()` on a column with nested values? See if you can work around this problem by using the `.split()` function and the DataFrame's [`.stack()` method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.stack.html). 


In [190]:
df_apps_clean.Genres.value_counts()

Tools                                718
Entertainment                        467
Education                            429
Productivity                         301
Personalization                      298
                                    ... 
Lifestyle;Pretend Play                 1
Racing;Pretend Play                    1
Adventure;Brain Games                  1
Communication;Creativity               1
Travel & Local;Action & Adventure      1
Name: Genres, Length: 114, dtype: int64

In [191]:
# Split the strings on the semi-colon and then .stack them.
stack = df_apps_clean.Genres.str.split(';', expand=True).stack()
print(f'We now have a single column with shape: {stack.shape}')

num_genres = stack.value_counts()
num_genres

We now have a single column with shape: (8564,)


Tools                      719
Education                  587
Entertainment              498
Action                     304
Productivity               301
Lifestyle                  298
Personalization            298
Finance                    296
Medical                    292
Sports                     270
Photography                263
Business                   262
Communication              258
Health & Fitness           245
Casual                     216
News & Magazines           204
Social                     203
Simulation                 200
Travel & Local             187
Arcade                     185
Shopping                   180
Books & Reference          171
Video Players & Editors    150
Dating                     134
Puzzle                     124
Maps & Navigation          118
Role Playing               111
Racing                     103
Action & Adventure          96
Strategy                    95
Food & Drink                94
Educational                 93
Adventur

# Colour Scales in Plotly Charts - Competition in Genres

**Challenge**: Can you create this chart with the Series containing the genre data? 

<img src=https://imgur.com/DbcoQli.png width=400>

Try experimenting with the built in colour scales in Plotly. You can find a full list [here](https://plotly.com/python/builtin-colorscales/). 

* Find a way to set the colour scale using the color_continuous_scale parameter. 
* Find a way to make the color axis disappear by using coloraxis_showscale. 

In [192]:
genres_bar = px.bar(x=num_genres.index[:15],
                    y=num_genres.values[:15],
                    title='Top Genres',
                    hover_name=num_genres.index[:15],
                    color=num_genres.values[:15],
                    color_continuous_scale='Agsunset_r') # set color scale

genres_bar.update_layout(xaxis_title='Genre',
                         yaxis_title='Number of Apps',
                         coloraxis_showscale=False) # turn off color axis

genres_bar.show()

# Grouped Bar Charts: Free vs. Paid Apps per Category

**Challenge**: Use the plotly express bar [chart examples](https://plotly.com/python/bar-charts/#bar-chart-with-sorted-or-ordered-categories) and the [.bar() API reference](https://plotly.com/python-api-reference/generated/plotly.express.bar.html#plotly.express.bar) to create this bar chart: 

<img src=https://imgur.com/LE0XCxA.png>

You'll want to use the `df_free_vs_paid` DataFrame that you created above that has the total number of free and paid apps per category. 

See if you can figure out how to get the look above by changing the `categoryorder` to 'total descending' as outlined in the documentation here [here](https://plotly.com/python/categorical-axes/#automatically-sorting-categories-by-name-or-total-value). 

In [193]:
df_apps_clean.Type.value_counts()

Free    7595
Paid     589
Name: Type, dtype: int64

In [194]:
# Step 1: Group our data first by Category and then by Type
# Step 2: Add up the number of apps per each type
df_free_vs_paid = df_apps_clean.groupby(["Category", "Type"], as_index=False).agg({'App': pd.Series.count})

df_free_vs_paid.sort_values('App')

Category  Type   App
3    AUTO_AND_VEHICLES  Paid     1
24      FOOD_AND_DRINK  Paid     2
38  NEWS_AND_MAGAZINES  Paid     2
40           PARENTING  Paid     2
17       ENTERTAINMENT  Paid     2
..                 ...   ...   ...
31           LIFESTYLE  Free   284
21             FINANCE  Free   289
53               TOOLS  Free   656
25                GAME  Free   834
19              FAMILY  Free  1456

[61 rows x 3 columns]

The key is using the `color` and `barmode` parameters for the `.bar()` method. To get a particular order, you can pass a dictionary to the axis parameter in `.update_layout()`.

In [195]:
free_vs_paid_bar = px.bar(df_free_vs_paid,
                          x='Category',
                          y='App',
                          title='Free vs Paid Apps by Category',
                          color='Type', # differentiate color by Type
                          barmode='group') # separate 2 types

free_vs_paid_bar.update_layout(xaxis_title='Category',
                               yaxis_title='Number of Apps',
                               xaxis={'categoryorder':'total descending'}, # set order
                               yaxis=dict(type='log'))

free_vs_paid_bar.show()

# Plotly Box Plots: Lost Downloads for Paid Apps

**Challenge**: Create a box plot that shows the number of Installs for free versus paid apps. How does the median number of installations compare? Is the difference large or small?

Use the [Box Plots Guide](https://plotly.com/python/box-plots/) and the [.box API reference](https://plotly.com/python-api-reference/generated/plotly.express.box.html) to create the following chart. 

<img src=https://imgur.com/uVsECT3.png>


In [196]:
box = px.box(df_apps_clean,
             x='Type',
             y='Installs')

box.update_layout(yaxis=dict(type='log'))

box.show()

In [197]:
type_box = px.box(df_apps_clean,
             x='Type',
             y='Installs',
             color='Type',
             notched=True, # difference of medians
             points='all', # display underlying data points
             title='How Many Downloads are Paid Apps Giving Up?')

type_box.update_layout(yaxis=dict(type='log'))

type_box.show()

Box plots show us some handy descriptive statistics in a graph - things like the median value, the maximum value, the minimum value, and some quartiles.

From the hover text in the chart, we see that the median number of downloads for free apps is 500,000, while the median number of downloads for paid apps is around 5,000! This is massively lower.

# Plotly Box Plots: Revenue by App Category

**Challenge**: See if you can generate the chart below: 

<img src=https://imgur.com/v4CiNqX.png>

Looking at the hover text, how much does the median app earn in the Tools category? If developing an Android app costs $30,000 or thereabouts, does the average photography app recoup its development costs?

Hint: I've used 'min ascending' to sort the categories. 

<img src="https://img-c.udemycdn.com/redactor/raw/2020-10-11_13-55-50-fba0062772e999739ca28119fd0cecda.png">

If an Android app costs 30,000USD to develop, then the average app in very few categories would cover that development cost. The median paid photography app earned about $20,000. Many more app’s revenues were even lower - meaning they would need other sources of revenue like advertising or in-app purchases to make up for their development costs. However, certain app categories seem to contain a large number of outliers that have much higher (estimated) revenue - for example in Medical, Personalisation, Tools, Game, and Family.

In [198]:
df_apps_clean.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size_MBs', 'Installs', 'Type',
       'Price', 'Content_Rating', 'Genres', 'Revenue_Estimate'],
      dtype='object')

In [199]:
df_paid_apps = df_apps_clean[df_apps_clean['Type'] == 'Paid']

box = px.box(df_paid_apps, 
             x='Category', 
             y='Revenue_Estimate',
             title='How Much Can Paid Apps Earn?')
 
box.update_layout(xaxis_title='Category',
                  yaxis_title='Paid App Ballpark Revenue',
                  xaxis={'categoryorder':'min ascending'},
                  yaxis=dict(type='log'))
 
 
box.show()

# How Much Can You Charge? Examine Paid App Pricing Strategies by Category

**Challenge**: What is the median price price for a paid app? Then compare pricing by category by creating another box plot. But this time examine the prices (instead of the revenue estimates) of the paid apps. I recommend using `{categoryorder':'max descending'}` to sort the categories.

In [200]:
df_paid_apps.Price.median()

2.99

In [201]:
box = px.box(df_paid_apps,
             x='Category',
             y="Price",
             title='Price per Category')
 
box.update_layout(xaxis_title='Category',
                  yaxis_title='Paid App Price',
                  xaxis={'categoryorder':'max descending'},
                  yaxis=dict(type='log'))
 
box.show()

The median price for an Android app is \$2.99.

However, some categories have higher median prices than others. This time we see that Medical apps have the most expensive apps as well as a median price of \$5.49. In contrast, Personalisation apps are quite cheap on average at \$1.49. Other categories which higher median prices are Business (\$4.99) and Dating (\$6.99). It seems like customers who shop in these categories are not so concerned about paying a bit extra for their apps.

<img src="https://img-c.udemycdn.com/redactor/raw/2020-10-11_13-55-50-fba0062772e999739ca28119fd0cecda.png">